expression/constraint_propagation.go

```diff
@@ -164,3 +172,132 @@ func ruleColumnEQConst(ctx sessionctx.Context, i, j int, exprs *exprSet) {
 		}
 	}
 }
+
+// ruleColumnLTConst propagates the "column > const" condition.
+func ruleColumnGTConst(ctx sessionctx.Context, i, j int, exprs *exprSet) {
+	ruleColumnXXConst(ctx, i, j, exprs, ast.GT, ast.LT, ast.GE)
+}
+
+// ruleColumnLTConst propagates the "column < const" condition.
+func ruleColumnLTConst(ctx sessionctx.Context, i, j int, exprs *exprSet) {
+	ruleColumnXXConst(ctx, i, j, exprs, ast.LT, ast.GT, ast.LE)
+}
+
+// ruleColumnXXConst propagates the "column OP const" condition, where op may be '>' and '<'
+func ruleColumnXXConst(ctx sessionctx.Context, i, j int, exprs *exprSet, GT string, LT string, GE string) {
+	cond := exprs.data[i]
+	f1, ok := cond.(*ScalarFunction)
+	if !ok || f1.FuncName.L != GT {
+		return
+	}
+	var col1 *Column
+	var con1 *Constant
+	col1, ok = f1.GetArgs()[0].(*Column)
+	if !ok {
+		return
+	}
+	con1, ok = f1.GetArgs()[1].(*Constant)
+	if !ok {
+		return
+	}
+
+	expr := exprs.data[j]
+	f2, ok := expr.(*ScalarFunction)
+	if !ok {
+		return
+	}
+
+	// col > c1, col > c2, c1 > c2 => col > c1
+	if f2.FuncName.L == GT {
+		col2, ok := f2.GetArgs()[0].(*Column)
+		if !ok || !col1.Equal(ctx, col2) {
+			return
+		}
+		con2, ok := f2.GetArgs()[1].(*Constant)
+		if !ok {
+			return
+		}
+		if !con1.RetType.Equal(con2.RetType) {
+			return
+		}
+
+		v, isNull, err := compareConstant(ctx, GT, con1, con2)
+		if err != nil {
+			log.Warn(err)
+			return
+		}
+		if !isNull && v > 0 {
+			exprs.tombstone[j] = true
+		}
+		return
+	}
+
+	// The simple case:
+	// col > c1, col < c2, c1 >= c2 => false
+	//
+	// The extended case:
+	// col > c1, f(col) < c2, f is monotonous, f(c1) <= c2 => false
+	//
+	// Prove:
+	// col > c1, f is monotonous => f(col) > f(c1)
+	// f(col) > f(c1), f(col) < c2, f(c1) >= c2 => false
+	if f2.FuncName.L == LT {
+		con2, ok := f2.GetArgs()[1].(*Constant)
+		if !ok {
+			return
+		}
+		arg0 := f2.GetArgs()[0]
+		// The simple case.
+		var fc1 Expression
+		col2, ok := arg0.(*Column)
+		if ok {
+			fc1 = con1
+		} else {
+			// The extended case.
+			scalarFunc, ok := arg0.(*ScalarFunction)
+			if !ok {
+				return
+			}
+			_, ok = monotoneFuncs[scalarFunc.FuncName.L]
+			if !ok {
+				return
+			}
+			col2, ok = scalarFunc.GetArgs()[0].(*Column)
+			if !ok {
+				return
+			}
+			var err error
+			fc1, err = NewFunction(ctx, scalarFunc.FuncName.L, scalarFunc.RetType, con1)
+			if err != nil {
+				log.Warn(err)
+				return
+			}
+		}
+		if !col1.Equal(ctx, col2) {
+			return
+		}
+		v, isNull, err := compareConstant(ctx, GE, fc1, con2)
+		if err != nil {
+			log.Warn(err)
+			return
+		}
+		if !isNull && v > 0 {
+			exprs.SetConstFalse()
+		}
+		return
+	}
+	return
+}
+
+var monotoneFuncs = map[string]struct{}{
+	ast.ToDays: {},
+}
+
+// compareConstant compares two expressions. c1 and c2 should be constant with the same type.
+func compareConstant(ctx sessionctx.Context, fn string, c1, c2 Expression) (int64, bool, error) {
+	cmp, err := NewFunction(ctx, fn, c1.GetType(), c1, c2)
```

Comment:
The third param of `NewFunction` is the return type of the new function. For binary comparison function. It should be `NewFieldType(mysql.TypeTiny)`.